### ISSUE TITLE
[Epic] Support bridging providers built with Terraform Plugin Framework

### ISSUE THREAD

**Issue by t0yv0** [pulumi/pulumi-terraform-bridge#744]
## Overview

Enable building Pulumi providers from Terraform providers that are defined fully or in part using the new [Terraform Plugin Framework](https://github.com/hashicorp/terraform-plugin-framework) by updating [pulumi-terraform-bridge](https://github.com/pulumi/pulumi-terraform-bridge/) to understand it.

The work in progress is available as [pf Go Module](https://github.com/pulumi/pulumi-terraform-bridge/tree/master/pf)
which includes a [README](https://github.com/pulumi/pulumi-terraform-bridge/tree/master/pf#readme) with instructions on
trying it out.

## Support

[#plugin-framework](https://pulumi-community.slack.com/archives/C04NPQTRKM4) in Community Slack.

## Timeline

2023 Q1

Releases:

- [v0.1.0](https://github.com/pulumi/pulumi-terraform-bridge/releases/tag/pf%2Fv0.1.0)
- [v0.2.0](https://github.com/pulumi/pulumi-terraform-bridge/releases/tag/pf%2Fv0.2.0)
- [v0.2.1](https://github.com/pulumi/pulumi-terraform-bridge/releases/tag/pf%2Fv0.2.1)
- [v0.3.0](https://github.com/pulumi/pulumi-terraform-bridge/releases/tag/pf%2Fv0.3.0)
- [v0.4.0](https://github.com/pulumi/pulumi-terraform-bridge/releases/tag/pf%2Fv0.4.0)
- [v0.5.0](https://github.com/pulumi/pulumi-terraform-bridge/releases/tag/pf%2Fv0.5.0)
- [v0.6.0](https://github.com/pulumi/pulumi-terraform-bridge/releases/tag/pf%2Fv0.6.0)
- [v0.6.1](https://github.com/pulumi/pulumi-terraform-bridge/releases/tag/pf%2Fv0.6.1)
- [v0.7.0](https://github.com/pulumi/pulumi-terraform-bridge/releases/tag/pf%2Fv0.7.0)

## Bridged Terraform Plugin Framework Providers ⚙️

Pulumi-maintained providers targeting an update soon:

- [pulumi-random](http://github.com/pulumi/pulumi-random)
- [pulumi-tls](http://github.com/pulumi/pulumi-tls)

Other provider ideas, open for community suggestions and unblocking issues encountered on community ports:

- github.com/fly-apps/terraform-provider-fly
- github.com/hashicorp/terraform-provider-tls
- github.com/aiven/terraform-provider-aiven
- github.com/civo/terraform-provider-civo
- github.com/fortinetdev/terraform-provider-fortios
- github.com/tmeckel/terraform-provider-mssql

## Work Items ⚙️

**M0.79**

- pulumi/home#2380
- pulumi/home#2387
- Investigate quick support options via muxing to v5 protocol

**M0.80**

- Read support
- Rewrote Diff/Create/Update with better understanding of state vs checkedInputs from Design syncs, remove TODOs
- Support more types on the wire
- Boolean
- List
- Map
- Support computing PackageSpec Pulumi schema from PF schema (by reusing tfgen)
- Resources
- Provider config
- Build new-style tfgen.Main
- Integration tests cover CRUD scenarios over random provider and test-only testbridge provider
- Unit tests replay gRPC logs from integration tests to check CRUD methods more quickly

**M0.81**

- Support Object types
- Ensure nested attributes work
- Ensure ID generation feature parity with prod tfbridge
- pulumi/home#2432
- pulumi/home#2452
  - pulumi/pulumi-terraform-bridge#656
  - pulumi/pulumi-terraform-bridge#657
  - pulumi/pulumi-terraform-bridge#663
- Ensure provider bridge honors all tfgen-initiated property renames
  - Currently only handling pluralized List property names
  - Need to handle anything tfgen decides to do
  - Need to handle user-initiated name overrides set in ProviderInfo
  - pulumi/pulumi-terraform-bridge#661
  - pulumi/pulumi-terraform-bridge#649

**M0.82**

- pulumi/home#2470
- Ensure data sources schema generation support
- pulumi/pulumi-terraform-bridge#697
- pulumi/pulumi-terraform-bridge#695

**M0.83**

- pulumi/home#2471
- pulumi/home#2472
- pulumi/home#2473

Review:

- pulumi/pulumi-terraform-bridge#666
- pulumi/pulumi-terraform-bridge#690
- pulumi/pulumi-terraform-bridge#694
- pulumi/pulumi-terraform-bridge#696
- pulumi/pulumi-terraform-bridge#698
- pulumi/pulumi-terraform-bridge#703
- pulumi/pulumi-terraform-bridge#707
- pulumi/pulumi-terraform-bridge#708
- pulumi/pulumi-terraform-bridge#715

**M0.84**

Dev:

- pulumi/pulumi-terraform-bridge#727 No Major.Minor.Patch
- pulumi/pulumi-terraform-bridge#729 PropertyValue should be secret but is not
- pulumi/home#2570
- pulumi/pulumi-terraform-bridge#758

Review:

 https://github.com/pulumi/home/issues/2543 PF 1.0

**M0.85**

- #730
- pulumi/pulumi-random#257
- pulumi/pulumi-terraform-bridge#748
- #751
- #797
- pulumi/pulumi-random#265
- pulumi/pulumi-random#279
- pulumi/pulumi-tls#173
- #872
- #732
- pulumi/pulumi-tls#149
- #812
- #795
- #749

**M0.86**

- [1/3] #820
- [1/3] pulumi/pulumi-tls#175
- [1/3] #819
- [1] #823
- [3] #828
- [3] #731
- [5] #830
- [3] #803
- [2] #750
- [1] #752
- [1] #821
- [1] #822
- [1] #753

**To Triage**

- #917
- #918
- #920
- #930
- #931
- #935
- #936
- #937
- #938

**Remaining**

- #824
- #827
- #794
- #754
- #747
- #793
- #829

**Comment by ryanrasti** [comment]
Wow -- what great timing! I was looking for this feature for writing our fly configuration with Pulumi (brought here by [this thread](https://community.fly.io/t/pulumi-provider/7931/5)).

Any idea on what the timeline for getting TPF support would be?

**Comment by t0yv0** [comment]
@ryanrasti it's my team's top priority for the quarter. What's available today is viable for small/simple providers and I'm working to itemize the list of known missing features and provides links to getting started. Several Pulumi partners are trying this out already and we will be dogfooding it in random and tls providers shortly. 

Any feedback you might have or even just upvoting issues is super helpful!

**Comment by tmeckel** [comment]
@t0yv0 Actually I wanted to publish the PGSoft MSSQL Terraform Provider as a Pulumi provider using TF Bridge. https://github.com/pulumiverse/infra/pull/59 But I wasn't aware of the fact that the provider is using the `terraform-plugin-framework`. So I suspect that I've to postpone my work until TF bridge will support `terraform-plugin-framework`. 

That's a small bummer because I planned to use the Pulumi version of the provider in a customer project to fulfill compliance requirements. Sad that I could upvote this issue here only once 😁

**Comment by tmeckel** [comment]
@t0yv0 Any guidance on how to use the [PF Version of TF Bridge](https://github.com/pulumi/pulumi-terraform-bridge/tree/master/pf)? I could try to add/integrate the required steps to wrap a TF provider using the `terraform-plugin-framework` in my [cookiecutter template](https://github.com/tmeckel/pulumi-tf-provider-cookiecutter) to create new Pulumi providers based on TF Bridge.

**Comment by t0yv0** [comment]
@tmeckel sorry for late reply here, I have been keeping quite busy! The cookiecutter template support sounds quite exciting actually.

The README you link to is the best we have for the moment, but we will have some example repositories shortly. Our current goal is to start using it ourselves as soon as practical. The first attempt to release pulumi-random encountered a critical issue with unexpected replace plans on `pf/v0.1.0`, which is fixed by #748, and now I'm working to try to upgrade `pulumi-random` again.

There're some trade-offs with trying to use pf/tfbridge before it reaches GA - it may not be suitable for all providers yet based on incomplete feature set support.  We try to keep the list in this issue up to date to help evaluate, and I'll start correlating it to versions as in a CHANGELOG shortly.

**Comment by t0yv0** [comment]
Sorry folks to ask for update the bookmarks, but according to our company process the remaining work rolls over to a continuation epic for Q2 2023: https://github.com/pulumi/pulumi-terraform-bridge/issues/956
